02-peer-protocol.md

```diff
@@ -430,6 +446,515 @@ would create a Denial of Service risk; therefore, forgetting it is recommended
+    - MUST send `funding_signed2`
+
+#### Rationale
+
+The first commitment transaction has no HTLC's in it.
+
+Note that the `commitment_signed` message will include the `channel_id` derived from
+the `funding_txid`, instead of the `temporary_node_id`.
+
+
+### The `funding_signed2` Message
+
+This message contains witness data for for the inputs that were
+originally exchanged in `funding_add_input`.
+
+1. type: 64 (`funding_signed2`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u16`:`num_witnesses`]
+    * [`num_witnesses*witness_stack`:`witness_stack`]
+
+1. subtype: `witness_stack`
+2. data:
+    * [`sha256`:`prevtx_txid`]
+    * [`u32`:`prevtx_vout`]
+    * [`u16`:`num_input_witness`]
+    * [`num_input_witness*witness_element`:`witness_element`]
+
+1. subtype: `witness_element`
+2. data:
+    * [`u16`:`len`]
+    * [`len*byte`:`witness`]
+
+#### Requirements
+The sending node:
+  - MUST verify it has received valid commitment signatures from its peer
+  - MUST set `witness` to the serialized witness data for the input
+    corresponding to `prevtx_txid`:`prevtx_vout`
+  - MUST remember the details of this funding transaction.
+  - MUST NOT send a `witness_stack` whose length exceeds its
+    previously indicated `max_witness_len`.
+  - if it has NOT received a valid `commitment_signed` message:
+    - MUST NOT send a `funding_signed2` message.
+
+The receiving node:
+  - if the `witness_stack` length exceeds `max_witness_len`:
+    - MUST error.
+  - MUST verify it has received valid commitment signatures
+  - SHOULD apply `witness`es to the funding transaction and
+    broadcast it.
+
+#### Rationale
+
+Every node must transmit their complete set of witness data for
+the inputs included in the funding transaction.
+
+`prevtx_txid` and `prevtx_vout` are the identifier for the input
+which the witness data pertain to
+
+`witness` is the data for a witness element in a witness stack
+
+### Kicking Off Replace-By-Fee: `init_rbf`
+
+This message initiates the flow to create a replacement funding
+transaction.  It is sent by the channel opener, after the funding
+transaction has been broadcast but before a `funding_locked` message
+is exchanged.
+
+Once an `init_rbf` message has been successfully ack'd by the accepter
+node, the message flow returns to `commitment_signed` and proceeds as
+indicated above.
+
+1. type: 62 (`init_rbf`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u64`:`funding_satoshis`]
+    * [`u32`:`feerate_per_kw`]
+    * [`u32`:`feerate_per_kw_funding`]
+    * [`u16`:`num_additional_inputs`]
+    * [`num_additional_inputs*input_info`:`input_info`]
+    * [`u16`:`num_outputs`]
+    * [`num_outputs*output_info`:`output_info`]
+
+#### Requirements
+
+The sending node:
+  - MUST have sent `open_channel2`.
+  - MUST send a `feerate_per_kw_funding` greater than the most recently
+    negotiated rate.
+  - MAY update the `feerate_per_kw`, the commitment transaction feerate.
+  - MAY include additional inputs.
```

Comment:
IIRC I think that hurts BIP 125 rule 2, you can't add new inputs to rbf'ed transactions

Reply:
This is a misinterpretation of rule 2; an "unconfirmed input" is a utxo that has not yet been confirmed in a block (mined). 

>2. The replacement transaction may only include an unconfirmed input if that input was included in one of the original transactions. (An unconfirmed input spends an output from a currently-unconfirmed transaction.)

I think this is an oversight of the spec though; it should strictly exclude unconfirmed inputs (and thus disallow chaining)

Reply:
You're right, thanks the rule is more liberal than I believed.

Yes sure the LN spec should disallow unconfirmed input to avoid chaining. That would be a real attack vector for zero-conf channels if people keep doing them. Maybe even ask for few confirmations to avoid messy 1-depth reorg edges cases.